RecoLocalTracker/SubCollectionProducers/src/SeedClusterRemoverPhase2.cc

```diff
@@ -0,0 +1,358 @@
+  }
+}
+
+
+SeedClusterRemoverPhase2::~SeedClusterRemoverPhase2()
+{
+}
+
+void SeedClusterRemoverPhase2::mergeOld(ClusterRemovalInfo::Indices &refs,
+                                            const ClusterRemovalInfo::Indices &oldRefs) 
+{
+        for (size_t i = 0, n = refs.size(); i < n; ++i) {
+            refs[i] = oldRefs[refs[i]];
+       }
+}
+
+ 
+template<typename T> 
+auto_ptr<edmNew::DetSetVector<T> >
+SeedClusterRemoverPhase2::cleanup(const edmNew::DetSetVector<T> &oldClusters, const std::vector<uint8_t> &isGood, 
+			     reco::ClusterRemovalInfo::Indices &refs, const reco::ClusterRemovalInfo::Indices *oldRefs){
+    typedef typename edmNew::DetSetVector<T>             DSV;
+    typedef typename edmNew::DetSetVector<T>::FastFiller DSF;
+    typedef typename edmNew::DetSet<T>                   DS;
+    auto_ptr<DSV> output(new DSV());
+    output->reserve(oldClusters.size(), oldClusters.dataSize());
+
+    unsigned int countOld=0;
+    unsigned int countNew=0;
+
+    // cluster removal loop
+    const T * firstOffset = & oldClusters.data().front();
+    for (typename DSV::const_iterator itdet = oldClusters.begin(), enddet = oldClusters.end(); itdet != enddet; ++itdet) {
+        DS oldDS = *itdet;
+
+        if (oldDS.empty()) continue; // skip empty detsets
+
+        uint32_t id = oldDS.detId();
+        DSF outds(*output, id);
+
+        for (typename DS::const_iterator it = oldDS.begin(), ed = oldDS.end(); it != ed; ++it) {
+            uint32_t index = ((&*it) - firstOffset);
+	    countOld++;
+            if (isGood[index]) { 
+                outds.push_back(*it);
+		countNew++;
+                refs.push_back(index); 
+                LogDebug("SeedClusterRemoverPhase2") << "SeedClusterRemoverPhase2::cleanup " << typeid(T).name() << " reference " << index << " to " << (refs.size() - 1);
+            }
+
+	}
+        if (outds.empty()) outds.abort(); // not write in an empty DSV
+    }
+
+    if (oldRefs != 0) mergeOld(refs, *oldRefs);
+    return output;
+}
+
+
+
+
+void SeedClusterRemoverPhase2::process(const TrackingRecHit *hit, float chi2, const TrackerGeometry* tg) {
+
+  DetId detid = hit->geographicalId(); 
+  uint32_t subdet = detid.subdetId();
+
+  assert(subdet > 0);
+
+  const type_info &hitType = typeid(*hit);
+  if (hitType == typeid(SiPixelRecHit)) {
+
+    if(!doPixel_) return;
+
+    const SiPixelRecHit *pixelHit = static_cast<const SiPixelRecHit *>(hit);
+    SiPixelRecHit::ClusterRef cluster = pixelHit->cluster();
+    LogDebug("SeedClusterRemoverPhase2") << "Plain Pixel RecHit in det " << detid.rawId();
+
+    if (cluster.id() != pixelSourceProdID) throw cms::Exception("Inconsistent Data") << 
+            "SeedClusterRemoverPhase2: pixel cluster ref from Product ID = " << cluster.id() << 
+            " does not match with source cluster collection (ID = " << pixelSourceProdID << ")\n.";
+
+    assert(cluster.id() == pixelSourceProdID);
```

Comment:
On 6/15/16 7:18 AM, Kevin Pedro wrote:
> In
> RecoLocalTracker/SubCollectionProducers/src/SeedClusterRemoverPhase2.cc
> <https://github.com/cms-sw/cmssw/pull/14865#discussion_r67169901>:
> 
>> +  assert(subdet > 0);
>> +
>> +  const type_info &hitType = typeid(*hit);
>> +  if (hitType == typeid(SiPixelRecHit)) {
>> +
>> +    if(!doPixel_) return;
>> +
>> +    const SiPixelRecHit *pixelHit = static_cast<const SiPixelRecHit *>(hit);
>> +    SiPixelRecHit::ClusterRef cluster = pixelHit->cluster();
>> +    LogDebug("SeedClusterRemoverPhase2") << "Plain Pixel RecHit in det " << detid.rawId();
>> +
>> +    if (cluster.id() != pixelSourceProdID) throw cms::Exception("Inconsistent Data") << 
>> +            "SeedClusterRemoverPhase2: pixel cluster ref from Product ID = " << cluster.id() << 
>> +            " does not match with source cluster collection (ID = " << pixelSourceProdID << ")\n.";
>> +
>> +    assert(cluster.id() == pixelSourceProdID);
> 
> It looks like this is also left over from the original algorithm, whose
> initial commit is described as " seed cluster remover: implementation
> sucks, and could be made less copy&paste, but no time now": 3794d2f
> <https://github.com/cms-sw/cmssw/commit/3794d2f703995ea3c4d7a00c8b1a33181048e59b>


A comment from 2012.
... and there is a hope that if we pass this copy-paste code now without
any further review,
it will get updated and improved soon.


> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/14865/files/1e83536fb64a8920c5d5ad1fc4ac6ed5f4927192#r67169901>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AEdcbhs-OivTu5xiVA30BKePTdDHBEmUks5qMAmigaJpZM4I0qFV>.
>